cmd/syft/cli/options/output.go

```diff
@@ -2,93 +2,209 @@ package options
-}
-
 var _ interface {
 	clio.FlagAdder
-} = (*MultiOutput)(nil)
+	clio.PostLoader
+} = (*Output)(nil)
+
+// Output has the standard output options syft accepts: multiple -o, --file, --template
+type Output struct {
+	AllowableOptions     []string `yaml:"-" json:"-" mapstructure:"-"`
+	AllowMultipleOutputs bool     `yaml:"-" json:"-" mapstructure:"-"`
+	Outputs              []string `yaml:"output" json:"output" mapstructure:"output"` // -o, the format to use for output
+	OutputFile           `yaml:",inline" json:"" mapstructure:",squash"`
+	OutputTemplate       `yaml:"template" json:"template" mapstructure:"template"`
+
+	// populated after configuration is loaded
+	encoders []sbom.FormatEncoder
+}
 
-func DefaultOutput() MultiOutput {
-	return MultiOutput{
-		Outputs: []string{string(table.ID)},
+func DefaultOutput() Output {
+	return Output{
+		AllowMultipleOutputs: true,
+		Outputs:              []string{string(table.ID)},
 	}
 }
 
-func (o *MultiOutput) AddFlags(flags clio.FlagSet) {
-	flags.StringArrayVarP(&o.Outputs, "output", "o",
-		fmt.Sprintf("report output format, options=%v", formats.AllIDs()))
+func (o *Output) AddFlags(flags clio.FlagSet) {
+	encs := format.DefaultEncoders()
+	var names []string
+	for _, e := range encs {
+		names = append(names, e.ID().String())
+	}
+	sort.Strings(names)
 
-	flags.StringVarP(&o.OutputTemplatePath, "template", "t",
-		"specify the path to a Go template file")
+	flags.StringArrayVarP(&o.Outputs, "output", "o",
+		fmt.Sprintf("report output format, options=%v", names))
 }
 
-func (o *MultiOutput) SBOMWriter() (sbom.Writer, error) {
-	return makeSBOMWriter(o.Outputs, o.File, o.OutputTemplatePath)
-}
+func (o *Output) PostLoad() error {
+	// setup all encoders based on the configuration
+	var list encoderList
+
+	// in the future there will be application configuration options that can be used to set the default output format
+	list.attempt(template.ID)(o.OutputTemplate.formatEncoders())
+	list.add(syftjson.ID)(syftjson.DefaultFormatEncoder())
+	list.add(table.ID)(table.DefaultFormatEncoder())
+	list.add(text.ID)(text.DefaultFormatEncoder())
+	list.add(github.ID)(github.DefaultFormatEncoder())
+	list.attempt(cyclonedxxml.ID)(cycloneDxXMLEncoders())
+	list.attempt(cyclonedxjson.ID)(cycloneDxJSONEncoders())
+	list.attempt(spdxjson.ID)(spdxJSONEncoders())
+	list.attempt(spdxtagvalue.ID)(spdxTagValueEncoders())
 
-// SingleOutput allows only 1 output to be specified, with a user able to set what options are allowed by setting AllowableOptions
-type SingleOutput struct {
-	AllowableOptions   []string `yaml:"-" json:"-" mapstructure:"-"`
-	Output             string   `yaml:"output" json:"output" mapstructure:"output"`
-	OutputTemplatePath string   `yaml:"output-template-path" json:"output-template-path" mapstructure:"output-template-path"` // -t template file to use for output
+	o.encoders = list.encoders
+
+	return list.err
 }
```

Comment:
Instead of a temporary variable on the `Output` struct and using this PostLoad function to populate the list of objects from the configuration, could it just be a helper function that's called when populating the internal configuration object? e.g. this function is something like:
```go
func (o *Output) FormatEncoders() (encoders []sbom.FormatEncoder, error) {
	// setup all encoders based on the configuration
	list := encoderList{}

	// in the future there will be application configuration options that can be used to set the default output format
	list.attempt(template.ID)(o.OutputTemplate.formatEncoders())
	list.add(syftjson.ID)(syftjson.DefaultFormatEncoder())
	list.add(table.ID)(table.DefaultFormatEncoder())
	list.add(text.ID)(text.DefaultFormatEncoder())
	list.add(github.ID)(github.DefaultFormatEncoder())
	list.attempt(cyclonedxxml.ID)(cycloneDxXMLEncoders())
	list.attempt(cyclonedxjson.ID)(cycloneDxJSONEncoders())
	list.attempt(spdxjson.ID)(spdxJSONEncoders())
	list.attempt(spdxtagvalue.ID)(spdxTagValueEncoders())
	
	return list.encoders, list.err
}
```

Reply:
yeah, that makes sense. I still want to keep it unexported but otherwise like it better with your suggestion